File: algoexpert/easy/two-number-sum/python/two-number-sum.py
# solution 1
# Time: O(n^2)
# Space: O(1)
def two_number_sum_sum_1(array, targetSum):
    for i in range(len(array) - 1):
        firstNum = array[i]
        for j in range(i + 1, len(array)):
            secondNum = array[j]
            if firstNum + secondNum == targetSum:
                return [firstNum, secondNum]
    return []

# solution 2
# Time: O(n)
# Space: O(n)
def two_number_sum_sum_2(array, targetSum):
    nums = {}
    for num in array:
        potentialMatch = targetSum - num
        if potentialMatch in nums:
            return [potentialMatch, num]
        else:
            nums[num] = True
    return []

# solution 3
def two_number_sum_sum_3(array, targetSum):
    array.sort()
    left = 0
    right = len(array) - 1
    while left < right:
        currentSum = array[left] + array[right]
        if currentSum == targetSum:
            return [array[left], array[right]]
        elif currentSum < targetSum:
            left += 1
        elif currentSum > targetSum:
            right -= 1
        else:
            return []

result1 = two_number_sum_sum_1([1,2,3,4,5,6,7,8,9,10], 10)
result2 = two_number_sum_sum_2([1,2,3,4,5,6,7,8,9,10], 10)
result3 = two_number_sum_sum_3([1,2,3,4,5,6,7,8,9,10], 10)
print(result1, result2, result3)

File: algoexpert/easy/two-number-sum/go/two-number-sum.go
package main

import (
	"fmt"
	"sort"
)

// solution1
func twoNumberSum1(array []int, targetSum int) []int {
	for i := 0; i < len(array)-1; i++ {
		firstNum := array[i]
		for j := 1; j < len(array); j++ {
			secondNum := array[j]
			if firstNum+secondNum == targetSum {
				return []int{firstNum, secondNum}
			}
		}
	}
	return []int{}
}

func twoNumberSum2(array []int, targetSum int) []int {
	nums := map[int]bool{}
	for _, num := range array {
		potentialMatch := targetSum - num
		if _, found := nums[potentialMatch]; found {
			return []int{potentialMatch, num}
		}
		nums[num] = true
	}
	return []int{}
}

func twoNumberSum3(array []int, targetSum int) []int {
	sort.Ints(array)
	left, right := 0, len(array)-1
	for left < right {
		currentSum := array[left] + array[right]
		if currentSum == targetSum {
			return []int{array[left], array[right]}
		} else if currentSum < targetSum {
			left += 1
		} else if currentSum > targetSum {
			right -= 1
		}
	}
	return []int{}
}

func main() {
	result1 := twoNumberSum1([]int{1, 2, 3, 4, 6, 7, 8, 9, 10}, 10)
	result2 := twoNumberSum2([]int{1, 2, 3, 4, 6, 7, 8, 9, 10}, 10)
	result3 := twoNumberSum3([]int{1, 2, 3, 4, 6, 7, 8, 9, 10}, 10)
	fmt.Println(result1, result2, result3)
}
